Add documentHighlighting to metals language server

tgodzik opened the pull request:
Works according to:
[https://microsoft.github.io/language-server-protocol/specification#textDocument_documentHighlight](documentHighlighting)

Let me know if I did something totally wrong, will need to add more tests still.

olafurpg committed:
Support navigation for library dependencies in Scala versions.

Previously, we only supported navigation inside library dependencies
with Scala version 2.12.8, after this commit it works with all Scala
versions including 2.11.x.

- Supported: definition, hover, signature help
- Not supported: completions, references

tgodzik committed:
Add documentHighlighting to metals LSP server

Previously, we didn't support documentHighlighting from metals lsp server and symbols were only highlighted based on their name.
Now we highlight only the same symbols. This includes highlighting both class and it's companion object, also work for parameters and vars.

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala:
```diff
@@ -53,6 +55,30 @@ final class DefinitionProvider(
+            }
+            .flatMap(_.range)
+            .map { range =>
+              new DocumentHighlight(range.toLSP)
```
This range may be outdated, we will need to use `TokenEditDistance` to adjust the tokens to the current buffer.

olafurpg replied:
See `ReferenceProvider` for an example

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala:
```diff
@@ -53,6 +55,30 @@ final class DefinitionProvider(
     }
   }
 
+  def symbolOccurrences(
```
Let's create a `DocumentHighlightProvider` class that has `DefinitionProvider` as a primary constructor parameter.

olafurpg reviewed (commented):
Very cool! Excited to try it out :)

olafurpg committed:
File watch project/project/*.sbt files

Previously, Metals didn't prompt to "Import changes" when editing
`project/project/plugins.sbt`. The reason for this behavior
was the assumption that `project/project/*.sbt` files didn't
affect the build sources but it turns out that is incorrect.

olafurpg committed:
Don't use `($0)` for implicit-only methods

olafurpg committed:
Remove authors and version from docstrings.

They have been more noise than helpful in my experience so far.

olafurpg committed:
Merge pull request #597 from olafurpg/flatten-snippet

Several small fixes

olafurpg committed:
Fix #599, don't insert import above generated parameter accessors

olafurpg committed:
Merge pull request #600 from olafurpg/flatten-snippet

Fix #599, don't insert import above generated parameter accessors

olafurpg committed:
Remove signature help fallback in hover.

Previously, hover used signature help as a fallback when the cursor was
over a named argument. After this commit, we show the method signature
instead. It would be nice to show only the parameter signature in the
future but this is commit is just to fix a bug that resulted in wrong
results.

olafurpg committed:
Handle desugaring of typechecked named arguments

olafurpg committed:
Mark frequently failing Windows tests as flaky.

olafurpg committed:
Merge pull request #602 from olafurpg/hover-param

Remove signature help fallback in hover.

olafurpg committed:
Complete filename when defining toplevel class/trait/object

Previously, Metals didn't provide completions when defining a class in
the toplevel of a file. Now we suggest a name matching the file.

olafurpg committed:
Merge pull request #603 from olafurpg/filename-complete

Complete filename when defining toplevel class/trait/object

olafurpg committed:
Eagerly load Scalafmt during initialized

olafurpg committed:
Don't eagerly load Scalafmt during testing.

This broke a few test cases. This commit also disables the Scalafmt
reporter until the first formatting request so that we don't spam the
user with error messages on startup.

olafurpg committed:
Eagerly load presentation compiler for open buffers.

Previously, we waited until the first completion/hover/signatureHelp
request to load the presentation compiler. After this commit, we load
the compiler eagerly and display a "slow task" status with timer so that
it's more clear that we're downloading artifacts from the internet.
Scalafmt also uses slow tasks now, for consistency.

olafurpg committed:
Merge pull request #604 from olafurpg/eager-scalafmt

Eagerly load Scalafmt during initialized

olafurpg committed:
Fix failing test cases.

Improves the fallback experience in non-vscode editors.

olafurpg committed:
Merge pull request #605 from olafurpg/eager-pc

Eagerly load presentation compiler for open buffers.

olafurpg committed:
Don't eagerly load presentation compiler during tests.

Previously, we always eagerly loaded the presentation compiler to
improve the out-of-the-box experience for new Metals users. However,
this behavior is undesirable for tests because it triggers a lot
of unnecessary computation on every test.

It's generally undesirable to customize the runtime for running tests
but in this particular case I feel it's justified.

olafurpg committed:
Handle exceptions when computing completion position.

Previously, we returned no completions in specific cases due to a null
pointer exception caused by a type error. Now we gracefully handle
type errors so that completions are still returned even when `Tree.tpe`
is null.

olafurpg committed:
Merge pull request #614 from olafurpg/load

Don't eagerly load presentation compiler during tests.

olafurpg committed:
Merge pull request #616 from olafurpg/npe

Handle exceptions when computing completion position.

olafurpg committed:
Fetch git tags before ci-release

olafurpg committed:
Improve completion snippets, fixes #610.

Previously, selecting a completion item would remove the trailing
identifier.

```scala
// before
List(1).m@@flatMap { x => Nil }
// after
List(1).map$0 { x => Nil }
```

This behavior was at best buggy and at worst surprising in cases where
it was working as intended.

Now, completions never strip the suffix unless the completed identifier
is identical to the suffix.

```scala
// before
object Foo extends Fut@@ure[Int]
// after
import scala.concurrent.Future
object Foo extends Future$0[Int]
```

Additionally, snippets like `($0)` and `[$0]` are now disabled
when editing inside an existing expression. For example,

```
// before
Syst@@println()
// old behavior
System$0()
// new behavior
System.$0println()
```

olafurpg committed:
Merge pull request #623 from olafurpg/conservative-snippets

Improve completion snippets, fixes #610.

olafurpg committed:
Disable macro-paradise compiler plugin, fixes #622.

Previously, we tried to expand macro annotations so that
completions/hover/signatureHelp would take the expanded code in the
current source file into account. This approach didn't work:

- the presentation compiler crashed frequently in macro-paradise
  on the second request for completions.
- completions/hover/signatureHelp don't work at all inside macro
  expanded code because macroparadise doesn't respect
  -Ymacro-expand:discard.

Now with macroparadise disabled, completions work as expected inside a
class with the `@react` annotation in https://github.com/shadaj/slinky.

Note that completions/hover/signatureHelp will continue to take macro
expanded code into account from other source files. It is only for
expanded code in the current open buffer where we won't take macro
expanded signatures into account.

mudsam committed:
Render @see as links. Improve inline link handling and update test cases.

olafurpg committed:
Merge pull request #625 from olafurpg/slinky

Disable macro-paradise compiler plugin, fixes #622.

mudsam committed:
Remove redundant pattern from markdown generator.

mudsam committed:
Fixed formatting

olafurpg committed:
Merge pull request #627 from mudsam/linkfix

Render @see as links. Improve inline link handling and update test cases.

olafurpg committed:
Sort symbols defined in the current file to the top, fixes #618.

Previously, we sorted global symbols defined in the current file in the
same group as global symbols from the root scope. Now, we priotize
symbols in the current file.

This change has the nice side-effect that `::` no longer appears at the
top of all completions in an empty scope without a prefix.

olafurpg committed:
Merge pull request #628 from olafurpg/local-sort

Sort symbols defined in the current file to the top, fixes #618.

olafurpg committed:
Provide fuzzier type member matching and more polished sorting.

Previously, the following completion returned empty results
```scala
class Foo { def getTimeStamp: Int = 42 }
new Foo().time@@
```
Now, the completion results include `getTimeStamp`.

Since we started allowing fuzzier results, we needed better sorting of
completion items in the same category. For example, in the following
example
```scala
class Foo {
  def toInt: Int = 42
  def instance: Int = 42
  def intNumber: Int = 42
}
new Foo().int@@
```
we want the results sorted in the following order
```scala
intNumber // because prefix match
toInt     // because literal substring match
instance  // last fallback
```

This improvement in sorting benefitted non-fuzzy non-type members too.
Previously, the query `Iterator` returned the order
`BufferedIterator,Iterator` because they were sorted alphabetically.
Now, the order is `Iterator,BufferedIterator` instead, a big
quality-of-life improvement.

olafurpg committed:
Make cross tests pass on 2.11

olafurpg committed:
Merge pull request #629 from olafurpg/local-sort

Provide fuzzier type member matching and more polished sorting.

olafurpg committed:
Fix local import position for case block bug, closes #617.

olafurpg committed:
Provide signature help for Java constructors, fixes #615.

Previously, we supported parameter hints for the expression
`new java.io.File(@@)` but not `new File(@@)`. This commit fixes that
bug.

The root cause was that we were manipulating the Java module symbol
instead of the class symbol.

olafurpg committed:
Use `class` instead of `object` hover keystring for Java classes, fixes #611.

olafurpg committed:
Add tests case to demonstrate bug fix, closes #608.

olafurpg committed:
Fix bug in `isKindaTheSameAs` method to compare symbols, fixes #607.

olafurpg committed:
Emit empty package in toplevel mtags, fixes #524

olafurpg committed:
Merge pull request #633 from olafurpg/bug-bonanza

Fix six bugs
